junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/MethodSourceUtils.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2015-2019 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * https://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junit.jupiter.engine.descriptor;
+
+import java.net.URI;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+import org.junit.platform.commons.util.Preconditions;
+import org.junit.platform.commons.util.ReflectionUtils;
+import org.junit.platform.engine.support.descriptor.MethodSource;
+
+/**
+ * @since 5.5
+ */
+class MethodSourceUtils {
+
+	static final String METHOD_SCHEME = "method";
+
+	static MethodSource fromUri(URI uri) {
+		Preconditions.notNull(uri, "URI must not be null");
+		Preconditions.condition(METHOD_SCHEME.equals(uri.getScheme()),
+			() -> "URI [" + uri + "] must have [" + METHOD_SCHEME + "] scheme");
+
+		Supplier<IllegalArgumentException> illegalArgumentExceptionSupplier = () -> new IllegalArgumentException(
+			"Invalid method URI");
+		String schemeSpecificPart = Optional.ofNullable(uri.getSchemeSpecificPart()).orElseThrow(
+			illegalArgumentExceptionSupplier);
+		String fragment = Optional.ofNullable(uri.getFragment()).orElseThrow(illegalArgumentExceptionSupplier);
```

Comment:
Please use `Preconditions.notNull` instead of `Optional.ofNullable().orElse()`.

Reply:
Sure.
I want also to improve Exception message by linking to [`DiscoverySelectors.selectMethod(String)`](https://junit.org/junit5/docs/current/api/org/junit/platform/engine/discovery/DiscoverySelectors.html#selectMethod(java.lang.String)) to help troubleshooting.
Something like `"Invalid method URI see https://junit.org/junit5/docs/5.4.1/api/org/junit/platform/engine/discovery/DiscoverySelectors.html#selectMethod(java.lang.String) for valid Fully Qualified Method Name"`

What do you think?

Reply:
I don't think we include any URLs in exception messages.

> Something like `"Invalid method URI see https://junit.org/junit5/docs/5.4.1/api/org/junit/platform/engine/discovery/DiscoverySelectors.html#selectMethod(java.lang.String) for valid Fully Qualified Method Name"`

How about this?

`"Invalid method URI. Consult the Javadoc for org.junit.platform.engine.discovery.DiscoverySelectors.selectMethod(String) for details on the supported format."`